src/security/KeyData.cc

```diff
@@ -86,51 +88,44 @@ Security::KeyData::loadX509ChainFromFile()
-        OPENSSL_free(nameStr);
-    } else
-#endif
-    {
-        debugs(83, DBG_PARSE_NOTE(3), "Using certificate chain in " << certFile);
-        // and add to the chain any other certificate exist in the file
-        CertPointer latestCert = cert;
-
-        while (const auto ca = Ssl::ReadX509Certificate(bio)) {
-            // get Issuer name of the cert for debug display
-            char *nameStr = X509_NAME_oneline(X509_get_subject_name(ca.get()), nullptr, 0);
-
-#if TLS_CHAIN_NO_SELFSIGNED // ignore self-signed certs in the chain
-            // self-signed certificates are not valid in a sent chain
-            if (X509_check_issued(ca.get(), ca.get()) == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(2), "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
-                OPENSSL_free(nameStr);
-                continue;
-            }
-#endif
-            // checks that the chained certs are actually part of a chain for validating cert
-            const auto checkCode = X509_check_issued(ca.get(), latestCert.get());
-            if (checkCode == X509_V_OK) {
-                debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << nameStr);
-                // OpenSSL API requires that we order certificates such that the
-                // chain can be appended directly into the on-wire traffic.
-                latestCert = CertPointer(ca);
-                chain.emplace_front(latestCert);
-            } else {
-                debugs(83, DBG_PARSE_NOTE(2), certFile << ": Ignoring non-issuer CA " << nameStr << ": " << X509_verify_cert_error_string(checkCode) << " (" << checkCode << ")");
+    debugs(83, DBG_PARSE_NOTE(3), "Using certificate chain in " << certFile);
+    // and add to the chain any other certificate exist in the file
+    CertList certsInFile;
+
+    while (const auto ca = Ssl::ReadX509Certificate(bio)) {
+        certsInFile.emplace_back(ca);
+    }
+
+    CertPointer latestCert = cert;
+    do {
+        CertPointer anIssuer;
+        // get the cert name for debug display
+        char *nameStr = X509_NAME_oneline(X509_get_subject_name(latestCert.get()), nullptr, 0);
+        // self-signed certificate, probably the root certificate, must not chained.
+        if (X509_check_issued(latestCert.get(), latestCert.get()) == X509_V_OK) {
+            debugs(83, DBG_PARSE_NOTE(2), "CA " << nameStr << " is self-signed, will not be chained: " << nameStr);
+        } else {
+            for (auto candidateIssuer: certsInFile) {
+                const auto checkCode = X509_check_issued(candidateIssuer.get(), latestCert.get());
+                if (checkCode == X509_V_OK) {
+                    // We found an issuer add it to the chain.
+                    debugs(83, DBG_PARSE_NOTE(3), "Adding issuer CA: " << nameStr);
+                    anIssuer = CertPointer(candidateIssuer);
+                    chain.emplace_front(anIssuer);
+                    break;
```

Comment:
If possible, please remove candidateIssuer from certsInFile before exiting the loop because the same certificate cannot be an issuer twice (unless it is a self-signed/root certificate and we do not chain those anyway). In other words, the same certificate cannot be a _successful_ candidateIssuer twice.

Reply:
My original thought on this was that because the list in most cases has about 1-2 certificates, the cost of removing the certificate is worst than accessing a list with more items.   However we do not care about speed in these function and moreover maybe it is possible to enter an infinity loop because of issuers issue each other.

commit f8726a45b

Reply:
> the cost of removing the certificate is worst than accessing a list with more items

> maybe it is possible to enter an infinity loop because of issuers issue each other

You are right, on both counts: The reason to remove the used entry is not to optimize performance but to ensure progress.